onnxscript/function_libs/tools/torch_prims/generate_prims_signatures.py

```diff
@@ -0,0 +1,341 @@
+            return_node = cg.TypingRefs.Optional(return_node)
+        return_nodes.append(return_node)
+    if len(return_nodes) == 1:
+        return return_nodes[0]
+    return cg.BuiltinTypeRef("tuple", *return_nodes)
+
+
+def format_arg_name(arg: torchgen.model.Argument) -> str:
+    """Returns the python compatible name of the given argument."""
+    if arg.name == "from":
+        return f"{arg.name}_"
+    return arg.name
+
+
+def create_signature(op_name: str, schema: FunctionSchema) -> cg.FunctionDef:
+    """Creates the signature for the given function."""
+    args = [
+        arg.argument if isinstance(arg, torchgen.model.SelfArgument) else arg
+        for arg in schema.arguments.positional
+    ]
+    kwargs = [
+        arg
+        for arg in schema.arguments.kwarg_only
+        if not isinstance(arg, torchgen.model.TensorOptionsArguments)
+    ]
+
+    py_args = [
+        cg.Arg(
+            format_arg_name(arg),
+            get_argument_type(arg),
+            default_value=cg.Constant(parse_default_value(arg))
+            if arg.default is not None
+            else None,
+        )
+        for arg in args
+    ]
+    if kwargs:
+        # Arguments after this point are keyword-only.
+        py_args += [
+            cg.Arg(
+                format_arg_name(kwarg),
+                get_argument_type(kwarg),
+                default_value=cg.Constant(parse_default_value(kwarg))
+                if kwarg.default is not None
+                else None,
+                is_kwarg=True,
+            )
+            for kwarg in kwargs
+        ]
+
+    return cg.FunctionDef(
+        op_name,
+        *py_args,
+        return_type=create_return_type(schema.returns),
+        body=[
+            cg.ThunkStmt(f'"""{schema}"""'),
+            cg.Raise(cg.Call(cg.Name("NotImplementedError"))),  # type: ignore[list-item]
+        ],
+    )
+
+
+def create_onnx_function_module(
+    functions: Sequence[tuple(str, FunctionSchema)],  # type: ignore[valid-type]
+) -> cg.Module:
+    """Creates the onnx function module."""
+    return cg.Module(
+        cg.ImportFrom("__future__", cg.Alias("annotations")),
+        *[
+            create_signature(name, schema)
+            for name, schema in functions
+            if should_generate_signature(name, schema)
+        ],
+    )
+
+
+def copyright_header() -> str:
+    """Creates the copyright header."""
+    dashline = f"# {'-' * 74}"
+    return textwrap.dedent(
+        f"""\
+        {dashline}
+        # Copyright (c) Microsoft Corporation. All rights reserved.
+        # Licensed under the MIT License.
+        {dashline}
+        # mypy: disable-error-code="misc,arg-type,type-arg,valid-type,assignment,return-value"
+        """
+    )
+
+
+def _get_func_schema_in_namespace(namespaces: List[_OpNamespace]) -> Dict[str, FunctionSchema]:
+    table: Dict[str, FunctionSchema] = {}
+    for op_namespace in namespaces:
+        for attr_name in dir(op_namespace):
```

Comment:
Not sure if torch.ops.prims has missing ops issue like we had torch.ops.aten (https://github.com/pytorch/pytorch/issues/99681). But reverse look up is generally preferred I think. Please take a look at https://github.com/pytorch/pytorch/pull/99682

Reply:
So far, there is no such problem for prims ops. But I agree we should find a way to guarantee the op list after all prims ops are stable. In the [doc](https://pytorch.org/docs/master/ir.html#prims-ir), pirms ops are still under development so we don't have a full list for them.